### PR TITLE
Pass +forced-atomics feature for riscv32{i,im,imc}-unknown-none-elf

### PR DESCRIPTION
**Pull request by taiki-e**
As said in https://github.com/rust-lang/rust/pull/98333#issuecomment-1666375293, `forced-atomics` target feature is also needed to enable atomic load/store on these targets (otherwise, libcalls are generated): https://godbolt.org/z/433qeG7vd

~~This PR is currently marked as a draft because:~~
- ~~`forced-atomics` target feature is currently broken (https://github.com/rust-lang/rust/issues/114153).~~ EDIT: Fixed
- ~~`forced-atomics` target feature has been added in LLVM 16 (https://github.com/llvm/llvm-project/commit/f5ed0cb217a9988f97b55f2ccb053bca7b41cc0c), but the current minimum LLVM version [is 15](https://github.com/rust-lang/rust/blob/90f0b24ad3e7fc0dc0e419c9da30d74629cd5736/src/bootstrap/llvm.rs#L557). In LLVM 15, the atomic load/store of these targets generates libcalls anyway.~~ EDIT: LLVM 15 has been dropped

Depending on the policy on the minimum LLVM version for these targets, this may be blocked until the minimum LLVM version is increased to 16.

r? @Amanieu

### CODE REVIEW

**Comment by MabezDev** [comment]
Just an FYI the check has moved to https://github.com/rust-lang/rust/blob/e1fcecb1b9930f109a8c1cf2b690e03483a41289/src/bootstrap/src/core/build_steps/llvm.rs#L567 and is still currently at a minimum version of 15.

**Comment by MabezDev** [comment]
> Depending on the policy on the minimum LLVM version for these targets, this may be blocked until the minimum LLVM version is increased to 16.

Is there a policy defined somewhere? If there isn't, in this case, it might be better to merge this now, as there is no downside to end users that I am aware of. 

We currently have quite a few issues moving to `portable-atomic` in the bare-metal esp riscv ecosystem because of these hardcoded defines: https://github.com/taiki-e/portable-atomic/blob/bc457f235b7dac1e64b4015f7c1b79d4897a8ae4/no_atomic.rs#L77-L79. They conflict with various atomic emulation methods we had been previously using with `atomic_polyfil`. We're trying to work around these, but I believe having this merged would fix most of them.

**Comment by taiki-e** [comment]
> hardcoded defines: https://github.com/taiki-e/portable-atomic/blob/bc457f235b7dac1e64b4015f7c1b79d4897a8ae4/no_atomic.rs#L77-L79.

The stuff about riscv, msp430, and avr on that list will be ignored, portable-atomic provides lock-free atomic load/store by default for those targets (to be precise: RISC-V with `not(target_has_atomic = "ptr")`, all msp430,  all avr) regardless of what is written there.

https://github.com/taiki-e/portable-atomic/blob/bc457f235b7dac1e64b4015f7c1b79d4897a8ae4/src/imp/riscv.rs

**Comment by MabezDev** [comment]
@taiki-e @Amanieu what are your thoughts on rebasing and getting this merged in its current state? All toolchains from rustup are shipping with LLVM16, older compilers built with LLVM15 will continue to work (without this feature). The only possible breakage is a custom build of modern rust + LLVM15, and even then that only applies to the riscv32 targets - seems like an unlikely source of breakage, with a huge upside for riscv32 developers.

**Comment by Dirbaio** [comment]
ping! 

Could we merge this? It's causing quite a bit of annoyances in the embedded ecosystem: porting crates to riscv32imc requires doing hacks to polyfill atomic loads/stores that the hardware is already capable of.

- Can we merge this before dropping LLVM 15? It seems LLVM silently ignores unknown target features, so code using atomics does compile with LLVM 15. It might generate broken code (with libcalls) but it doesn't break *more* code, since code using atomics simply didn't compile at all before this PR, so it's a net improvement, no regressions.
- If not, can we drop LLVM 15 now? What's the policy?

**Comment by taiki-e** [comment]
Rebased.

I would like to merge this if there is no problem with the policy on the minimum LLVM version for these targets, but I'm not a maintainer, so I don't know about that.

> * It might generate broken code (with libcalls) but it doesn't break _more_ code, since code using atomics simply didn't compile at all before this PR, so it's a net improvement, no regressions.

No, this causes a regression with LLVM 15 because compiler-builtins (and maybe libcore too) has code that depends on cfg(target_has_atomic_load_store). This is an error that actually occurred when https://github.com/rust-lang/rust/pull/98333 (already reverted) was merged: https://github.com/taiki-e/semihosting/commit/472e58a1ddbea9d546e658f929e961929059fe82

**Comment by Dirbaio** [comment]
https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/riscv.20forced-atomics drops llvm 15.

**Comment by MabezDev** [comment]
https://github.com/rust-lang/rust/pull/117947 has been merged, which unblocks this PR.

File: compiler/rustc_codegen_llvm/src/llvm_util.rs [hunk 1]
```diff
@@ -507,6 +507,8 @@ pub(crate) fn global_llvm_features(sess: &Session, diagnostics: bool) -> Vec<Str
             .features
             .split(',')
             .filter(|v| !v.is_empty() && backend_feature_name(v).is_some())
+            // Drop +forced-atomics feature introduced in LLVM 16.
+            .filter(|v| *v != "+forced-atomics" || get_version() >= (16, 0, 0))
```
Comment (Dirbaio) [inline compiler/rustc_codegen_llvm/src/llvm_util.rs:511]: this can be removed, LLVM16 is required now.

**Comment by taiki-e** [comment]
Rebased (and removed the no longer needed LLVM 15 related code).

**Comment by Dirbaio** [comment]
@rustbot label +S-waiting-on-review -S-blocked

ping @Amanieu

**Comment by Dirbaio** [comment]
instant review! :zap:

**Comment by taiki-e** [comment]
```
  Step 1/18 : FROM ubuntu:22.04
  error parsing HTTP 429 response body: invalid character 'T' looking for beginning of value: "Too Many Requests (HAP429).\n"
```
It seems to be a docker-related network error (unrelated to this PR's change)

**Comment by MabezDev** [comment]
Still network related by the looks of it :/

**Pull request merged**